### PR TITLE
Fix MetadataValidator

### PR DESCRIPTION
**Pull request by bernt-matthias**
If no `check` or `skip` attribute is given, `self.check` and `self.skip` need to be initialized to `None`. Otherwise validation with `value.missing_meta(check=self.check, skip=self.skip)` is called with `value.missing_meta(check=[""], skip=[""])`.

This led to the stalling observed in https://github.com/galaxyproject/galaxy/issues/13136

Explanation: each data input parameter automatically gets a `MetadataValidator`. Since this validator was dysfunctional the job was scheduled anyway, but did not execute.

Appears that the validator might have been dysfunctional for a long time (since `check` and `skip` have been introduced here:  https://github.com/bernt-matthias/galaxy/commit/b9e242e4eb39f7d2d582008affc7ec694cbe170b)

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

**Comment by mvdbeek** [comment]
Can you add a test here ?

**Comment by bernt-matthias** [comment]
> Can you add a test here ?

Also thought that this would be a good idea. Will try.

File: lib/galaxy/tools/parameters/validation.py [hunk 1]
```diff
@@ -561,18 +561,24 @@ class MetadataValidator(Validator):
...
-        self.check = check.split(",")
-        self.skip = skip.split(",")
+        if check == "":
+            self.check = None
+        else:
+            self.check = check.split(",")
+        if skip == "":
+            self.skip = None
+        else:
+            self.skip = skip.split(",")
```
Comment (nsoranzo) [inline lib/galaxy/tools/parameters/validation.py]:
```suggestion
        self.check = check.split(",") if check else None
        self.skip = skip.split(",") if skip else None
```

File: lib/galaxy/tools/parameters/validation.py [hunk 2]
```diff
@@ -561,18 +561,24 @@ class MetadataValidator(Validator):
...
-        if not message:
-            # TODO message not useful for negate="true" .. but maybe OK since the validator itself is not useful then
-            message = "Metadata missing, click the pencil icon in the history item to edit / save the metadata attributes"
         return cls(message=message,
                    check=elem.get('check', ""),
                    skip=elem.get('skip', ""),
                    negate=elem.get('negate', 'false'))
 
     def __init__(self, message=None, check="", skip="", negate='false'):
+        if message is None:
```
Comment (nsoranzo) [inline lib/galaxy/tools/parameters/validation.py]:
```suggestion
        if not message:
```

**Comment by nsoranzo** [comment]
You can replace all the ` no_validation="true"` with a better fix:
```diff
diff --git a/lib/galaxy/datatypes/data.py b/lib/galaxy/datatypes/data.py
index f7bc205d94..f3862a60e4 100644
--- a/lib/galaxy/datatypes/data.py
+++ b/lib/galaxy/datatypes/data.py
@@ -240,7 +240,7 @@ class Data(metaclass=DataMeta):
         for key, value in to_check:
             if key in skip or (not check and dataset.metadata.spec[key].get("optional")):
                 continue  # we skip check for optional and nonrequested values here
-            if not value:
+            if value is None:
                 return True
         return False
 
```

Values like 0, `""` or `[]` for a metadata should be valid.

**Comment by bernt-matthias** [comment]
> You can replace all the no_validation="true" with a better fix:

Yes :) I just realized this as well. Not sure yet if I will keep it anyway, because removing the default metadata validator makes debugging the tests a bit easier .. 

> Values like 0, "" or [] for a metadata should be valid.

Hah, just stumbled over the same point. Intuitively I thought that we should check if the value is the `no_value` provided when initializing the `MetadataElement`. But it seems that this is not useful in some cases. For instance `TabularData` has `comment_lines` with `no_value="0"`, i.e. we can not distinguish the cases where the `comment_lines` metadata has not been set from the case where it has been set to `0`.

**Comment by nsoranzo** [comment]
> > You can replace all the no_validation="true" with a better fix:
> 
> Yes :) I just realized this as well. Not sure yet if I will keep it anyway, because removing the default metadata validator makes debugging the tests a bit easier ..

I'd not use `no_validation` (apart from the fact that it has a misleading name, it should be `no_metadata_validation`) so that issues like these are not missed.

File: lib/galaxy/tools/parameters/validation.py [hunk 3]
```diff
@@ -18,7 +18,7 @@
 def get_test_fname(fname):
     """Returns test data filename"""
     path, name = os.path.split(__file__)
-    full_path = os.path.join(path, 'test', fname)
+    full_path = os.path.join(path, '..', '..', '..', '..', 'test-data', fname)
```
Comment (nsoranzo) [inline lib/galaxy/tools/parameters/validation.py]: This breaks the package structure, and therefore the `test_galaxy_packages` tests.

Comment (bernt-matthias) [reply lib/galaxy/tools/parameters/validation.py]: True. What would be the simplest solution for this? Create a test data directory in `tools` and link the test data?

Comment (nsoranzo) [reply lib/galaxy/tools/parameters/validation.py]: There's already one, with symlinks: lib/galaxy/tools/parameters/test/

**Comment by bernt-matthias** [comment]
> Values like 0, "" or [] for a metadata should be valid.

Spent quite a bit of time thinking about this. While formally correct the statement is not specific enough to ensure if a metadata element is set or not. In my test the tool did still stall, i.e. the failing `set_meta` in the upload job was not detected in the subsequent step.

I think the it might be the best to check if the value is equal to `no_value` that has been introduced here: https://github.com/galaxyproject/galaxy/commit/584999c5f0da243c2a936406e7034b7e2b39e02b

Unfortunately, apart from the commit message of the original commit `no_value` has not been documented and seems to be misused in some/many datatypes, e.g. there are metadata that have identical `no_value` and `default` which makes it impossible to check if metadata has been set. I estimate that there are approx 180 metadata elements in the sources that we would need to check... 

Essentially we should "ensure" (at least in the datatype definitions) that `no_value` is a value that can not be set by the `set_meta` function of the corresponding data type (we could be really strict and check this in the setter or relaxed and just fix the metdata definitions of the data types). But I'm not sure about valid combinations of `default`, `no_value`, and `optional`.

Besides this I was wondering if the test `meta_val != meta_spec.no_value` in `element_is_set` is correct given that both values can be `None`. 

I was also wondering if it wouldn't be better to fail the upload job if metadata setting fails.

File: lib/galaxy/datatypes/data.py [hunk 4]
```diff
@@ -229,18 +229,18 @@ def missing_meta(self, dataset, check=None, skip=None):
...
-        else:
-            to_check = dataset.metadata.items()
-        for key, value in to_check:
-            if key in skip or (not check and dataset.metadata.spec[key].get("optional")):
+        if not check:
+            check = dataset.metadata.keys()
+        for key in check:
+            if key in skip:
+                continue
+            if len(check) == 0 and len(skip) == 0 and dataset.metadata.spec[key].get("optional"):
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/data.py]: Now that you are reusing `check`, `len(check)` will never be 0 inside the loop.

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 5]
```diff
@@ -3730,7 +3730,10 @@ In case of ``data_collection`` parameters and
 separately for each of the contained data sets. Note that, for ``data``
 parameters a ``metadata`` validator is added automatically.
 
-- ``metadata``: Check for missing metadata.
+- ``metadata``: Check for missing metadata. The set of (optional) metadata
+to be checked can be set using either the ``check`` or ``skip`` attribute.
+Note that each data parameter has automatically a metadata validator that checks
+if all non-optional metadata is set, i.e. ``<validator type="metadata/>``.
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
if all non-optional metadata are set, i.e. ``<validator type="metadata/>``.
```

**Review by mvdbeek — APPROVED** [review]
Is this ready ?

**Comment by bernt-matthias** [comment]
With an approval of @nsoranzo yes :)

**Pull request merged**